java/client/src/org/openqa/selenium/support/ui/ExpectedConditions.java

```diff
@@ -933,4 +934,613 @@ private static WebElement findElement(By by, WebDriver driver) {
+        return String.format("value to contain \"%s\". Current value: \"%s\"", value, currentValue);
+      }
+    };
+  }
+
+  /**
+   * An expectation for checking WebElement with given locator has attribute which contains specific
+   * value
+   *
+   * @param locator   used to define WebElement to check it's parameters
+   * @param attribute used to define css or html attribute
+   * @param value     used as expected attribute value
+   * @return Boolean true when element has css or html attribute which contains the value
+   */
+  public static ExpectedCondition<Boolean> attributeContains(final By locator,
+                                                             final String attribute,
+                                                             final String value) {
+    return new ExpectedCondition<Boolean>() {
+      private String currentValue = null;
+
+      @Override
+      public Boolean apply(WebDriver driver) {
+        Boolean contains = false;
+        try {
+          currentValue = driver.findElement(locator).getAttribute(attribute);
+          if (currentValue == null) {
+            currentValue = driver.findElement(locator).getCssValue(attribute);
+          }
+          contains = currentValue.contains(value);
+        } catch (Exception e) {/**/}
+        return contains;
+      }
+
+      @Override
+      public String toString() {
+        return String
+          .format("value to contain \"%s\". Current value: \"%s\"", value, currentValue);
+      }
+    };
+  }
+
+  /**
+   * An expectation for checking WebElement any non empty value for given attribute
+   *
+   * @param element   used to check it's parameters
+   * @param attribute used to define css or html attribute
+   * @return Boolean true when element has css or html attribute with non empty value
+   */
+  public static ExpectedCondition<Boolean> attributeNotEmpty(final WebElement element,
+                                                             final String attribute) {
+    return new ExpectedCondition<Boolean>() {
+      private String currentValue = null;
+
+      @Override
+      public Boolean apply(WebDriver driver) {
+        Boolean hasText = false;
+        try {
+          currentValue = element.getAttribute(attribute);
+          if (currentValue == null) {
+            currentValue = element.getCssValue(attribute);
+          }
+          hasText = !currentValue.isEmpty();
+        } catch (Exception e) {/**/}
+        return hasText;
+      }
+    };
+  }
+
+  /**
+   * An expectation for checking child WebElement as a part of parent element to be visible
+   *
+   * @param locator     used to check parent element. For example table with locator
+   *                    By.xpath("//table")
+   * @param sub_locator used to find child element. For example td By.xpath("./tr/td")
+   * @return visible subelement
+   */
+  public static ExpectedCondition<WebElement> visibilityOfSubElementLocatedBy(final By locator,
```

Comment:
not sure I agree with the premise of this method.  we could end up in a very deep nesting of locations, e.g.: `visibilityOfSubSubSubSubElementLocatedBy(By,By,By,By,By)`

I think it'd be better just use `visibilityOfElementLocatedBy`